File: evals/promptfoo/assertions.py
from __future__ import annotations

import re
from typing import Any


SECTION_NAMES = {
    "chosen_subskills",
    "skill_file_reads",
    "routing_context",
    "plan",
    "execution",
    "validation",
}


def _parse_sections(output: str) -> dict[str, str]:
    sections: dict[str, list[str]] = {}
    current: str | None = None

    for line in output.splitlines():
        stripped = line.strip()
        if stripped in SECTION_NAMES:
            current = stripped
            sections.setdefault(current, [])
            continue
        if current:
            sections[current].append(line)

    return {name: "\n".join(lines).strip() for name, lines in sections.items()}


def _extract_skills(block: str) -> list[str]:
    skills: list[str] = []
    for raw_line in block.splitlines():
        line = raw_line.strip()
        if not line.startswith("-"):
            continue
        if line.startswith("- [provenance:"):
            continue

        code_match = re.search(r"`([^`]+)`", line)
        if code_match:
            skills.append(code_match.group(1))
            continue

        plain = line[1:].strip()
        if ":" in plain:
            plain = plain.split(":", 1)[0].strip()
        if plain:
            skills.append(plain)

    return skills


def _extract_file_reads(block: str) -> list[str]:
    reads: list[str] = []
    for raw_line in block.splitlines():
        line = raw_line.strip()
        if line.startswith("- "):
            reads.append(line[2:].strip())
    return reads


def _strip_wrapping_backticks(value: str) -> str:
    if value.startswith("`") and value.endswith("`") and len(value) >= 2:
        return value[1:-1]
    return value


def contract_assert(output: str, context: dict[str, Any]) -> dict[str, Any]:
    config = (context or {}).get("config", {})
    required_sections = config.get("required_sections", ["chosen_subskills", "skill_file_reads", "plan"])
    sections = _parse_sections(output)
    missing = [name for name in required_sections if not sections.get(name)]

    if missing:
        return {
            "pass": False,
            "score": 0,
            "reason": f"Missing required sections: {', '.join(missing)}",
        }

    return {"pass": True, "score": 1}


def expected_skill_set_assert(output: str, context: dict[str, Any]) -> dict[str, Any]:
    config = (context or {}).get("config", {})
    expected = list(config.get("expected_skills", []))
    exact_order = bool(config.get("exact_order", False))
    sections = _parse_sections(output)
    actual = _extract_skills(sections.get("chosen_subskills", ""))

    if exact_order:
        if actual[: len(expected)] != expected:
            return {
                "pass": False,
                "score": 0,
                "reason": f"Expected chosen_subskills to start with {expected}, got {actual}",
            }
    else:
        missing = [skill for skill in expected if skill not in actual]
        if missing:
            return {
                "pass": False,
                "score": 0,
                "reason": f"Missing expected skills: {missing}; got {actual}",
            }

    return {"pass": True, "score": 1}


def expected_any_skill_assert(output: str, context: dict[str, Any]) -> dict[str, Any]:
    config = (context or {}).get("config", {})
    expected_any = list(config.get("expected_any_skills", []))
    sections = _parse_sections(output)
    actual = _extract_skills(sections.get("chosen_subskills", ""))

    if not any(skill in actual for skill in expected_any):
        return {
            "pass": False,
            "score": 0,
            "reason": f"Expected at least one of {expected_any}, got {actual}",
        }

    return {"pass": True, "score": 1}


def no_disallowed_skills_assert(output: str, context: dict[str, Any]) -> dict[str, Any]:
    config = (context or {}).get("config", {})
    disallowed = set(config.get("disallowed_skills", []))
    sections = _parse_sections(output)
    actual = _extract_skills(sections.get("chosen_subskills", ""))
    found = [skill for skill in actual if skill in disallowed]

    if found:
        return {
            "pass": False,
            "score": 0,
            "reason": f"Found disallowed skills in chosen_subskills: {found}",
        }

    return {"pass": True, "score": 1}


def skill_file_reads_cover_selected_assert(output: str, context: dict[str, Any]) -> dict[str, Any]:
    config = (context or {}).get("config", {})
    managed_skills = set(config.get("managed_skills", []))
    sections = _parse_sections(output)
    selected = _extract_skills(sections.get("chosen_subskills", ""))
    reads = _extract_file_reads(sections.get("skill_file_reads", ""))

    missing: list[str] = []
    for skill in selected:
        if managed_skills and skill not in managed_skills:
            continue
        target = f"/{skill}/SKILL.md"
        if not any(target in path or path.endswith(f"{skill}/SKILL.md") for path in reads):
            missing.append(skill)

    if missing:
        return {
            "pass": False,
            "score": 0,
            "reason": f"Missing skill_file_reads coverage for selected managed skills: {missing}; reads={reads}",
        }

    return {"pass": True, "score": 1}


def skill_file_reads_only_child_skills_assert(output: str, context: dict[str, Any]) -> dict[str, Any]:
    sections = _parse_sections(output)
    reads = [_strip_wrapping_backticks(path) for path in _extract_file_reads(sections.get("skill_file_reads", ""))]

    invalid: list[str] = []
    for path in reads:
        if not path.endswith("/SKILL.md"):
            invalid.append(path)
            continue
        if "/skill-suite-orchestrator/SKILL.md" in path:
            invalid.append(path)

    if invalid:
        return {
            "pass": False,
            "score": 0,
            "reason": f"skill_file_reads must list only child SKILL.md files; found invalid entries: {invalid}",
        }

    return {"pass": True, "score": 1}


def mutually_exclusive_skills_assert(output: str, context: dict[str, Any]) -> dict[str, Any]:
    config = (context or {}).get("config", {})
    groups = list(config.get("mutually_exclusive_groups", []))
    sections = _parse_sections(output)
    actual = set(_extract_skills(sections.get("chosen_subskills", "")))

    violations: list[list[str]] = []
    for group in groups:
        present = [skill for skill in group if skill in actual]
        if len(present) > 1:
            violations.append(present)

    if violations:
        return {
            "pass": False,
            "score": 0,
            "reason": f"Found mutually exclusive skills together: {violations}",
        }

    return {"pass": True, "score": 1}

File: evals/promptfoo/assertions.mjs
const SECTION_NAMES = new Set([
  "chosen_subskills",
  "skill_file_reads",
  "routing_context",
  "plan",
  "execution",
  "validation",
  "telemetry",
]);

function parseSections(output) {
  const sections = new Map();
  let current = null;

  for (const line of output.split("\n")) {
    const stripped = line.trim();
    if (SECTION_NAMES.has(stripped)) {
      current = stripped;
      if (!sections.has(current)) {
        sections.set(current, []);
      }
      continue;
    }

    if (current) {
      sections.get(current).push(line);
    }
  }

  return Object.fromEntries(
    Array.from(sections.entries(), ([name, lines]) => [name, lines.join("\n").trim()]),
  );
}

function extractSkills(block) {
  const skills = [];

  for (const rawLine of block.split("\n")) {
    const line = rawLine.trim();
    if (!line.startsWith("-")) {
      continue;
    }
    if (line.startsWith("- [provenance:")) {
      continue;
    }

    const content = line.slice(1).trim();
    const leadMatch = content.match(/^(?:`([^`]+)`|([A-Za-z0-9:_-]+))(?=\s*[:\-]|\s*$)/);
    if (leadMatch) {
      skills.push(leadMatch[1] ?? leadMatch[2]);
      continue;
    }

    let plain = content;
    if (plain.includes(":")) {
      plain = plain.split(":", 1)[0].trim();
    }
    if (plain) {
      skills.push(plain);
    }
  }

  return skills;
}

function extractFileReads(block) {
  const reads = [];
  for (const rawLine of block.split("\n")) {
    const line = rawLine.trim();
    if (line.startsWith("- ")) {
      reads.push(stripWrappingBackticks(line.slice(2).trim()));
    }
  }
  return reads;
}

function stripWrappingBackticks(value) {
  if (value.startsWith("`") && value.endsWith("`") && value.length >= 2) {
    return value.slice(1, -1);
  }
  return value;
}

export function contractAssert(output, context) {
  const requiredSections = context?.config?.required_sections ?? [
    "chosen_subskills",
    "skill_file_reads",
    "routing_context",
    "plan",
    "execution",
    "validation",
    "telemetry",
  ];
  const sections = parseSections(output);
  const missing = requiredSections.filter((name) => !sections[name]);

  if (missing.length) {
    return {
      pass: false,
      score: 0,
      reason: `Missing required sections: ${missing.join(", ")}`,
    };
  }

  return { pass: true, score: 1, reason: "Required sections are present." };
}

export function expectedSkillSetAssert(output, context) {
  const expected = context?.config?.expected_skills ?? [];
  const exactOrder = Boolean(context?.config?.exact_order);
  const sections = parseSections(output);
  const actual = extractSkills(sections.chosen_subskills ?? "");

  if (exactOrder) {
    const actualPrefix = actual.slice(0, expected.length);
    const matches =
      actualPrefix.length === expected.length &&
      actualPrefix.every((skill, index) => skill === expected[index]);

    if (!matches) {
      return {
        pass: false,
        score: 0,
        reason: `Expected chosen_subskills to start with ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`,
      };
    }
  } else {
    const missing = expected.filter((skill) => !actual.includes(skill));
    if (missing.length) {
      return {
        pass: false,
        score: 0,
        reason: `Missing expected skills: ${JSON.stringify(missing)}; got ${JSON.stringify(actual)}`,
      };
    }
  }

  return { pass: true, score: 1, reason: "Expected skill set matched." };
}

export function expectedAnySkillAssert(output, context) {
  const expectedAny = context?.config?.expected_any_skills ?? [];
  const sections = parseSections(output);
  const actual = extractSkills(sections.chosen_subskills ?? "");

  if (!expectedAny.some((skill) => actual.includes(skill))) {
    return {
      pass: false,
      score: 0,
      reason: `Expected at least one of ${JSON.stringify(expectedAny)}, got ${JSON.stringify(actual)}`,
    };
  }

  return { pass: true, score: 1, reason: "At least one expected skill was selected." };
}

export function noDisallowedSkillsAssert(output, context) {
  const disallowed = new Set(context?.config?.disallowed_skills ?? []);
  const sections = parseSections(output);
  const actual = extractSkills(sections.chosen_subskills ?? "");
  const found = actual.filter((skill) => disallowed.has(skill));

  if (found.length) {
    return {
      pass: false,
      score: 0,
      reason: `Found disallowed skills in chosen_subskills: ${JSON.stringify(found)}`,
    };
  }

  return { pass: true, score: 1, reason: "No disallowed skills were selected." };
}

export function skillFileReadsCoverSelectedAssert(output, context) {
  const managedSkills = new Set(context?.config?.managed_skills ?? []);
  const sections = parseSections(output);
  const selected = extractSkills(sections.chosen_subskills ?? "");
  const reads = extractFileReads(sections.skill_file_reads ?? "");

  const missing = [];
  for (const skill of selected) {
    if (managedSkills.size && !managedSkills.has(skill)) {
      continue;
    }

    const target = `/${skill}/SKILL.md`;
    const covered = reads.some((path) => path.includes(target) || path.endsWith(`${skill}/SKILL.md`));
    if (!covered) {
      missing.push(skill);
    }
  }

  if (missing.length) {
    return {
      pass: false,
      score: 0,
      reason: `Missing skill_file_reads coverage for selected managed skills: ${JSON.stringify(missing)}; reads=${JSON.stringify(reads)}`,
    };
  }

  return { pass: true, score: 1, reason: "Every selected managed skill has a matching child SKILL.md read." };
}

export function skillFileReadsOnlyChildSkillsAssert(output) {
  const sections = parseSections(output);
  const reads = extractFileReads(sections.skill_file_reads ?? "");
  const invalid = [];

  for (const path of reads) {
    if (!path.endsWith("/SKILL.md")) {
      invalid.push(path);
      continue;
    }
    if (path.includes("/skill-suite-orchestrator/SKILL.md")) {
      invalid.push(path);
    }
  }

  if (invalid.length) {
    return {
      pass: false,
      score: 0,
      reason: `skill_file_reads must list only child SKILL.md files; found invalid entries: ${JSON.stringify(invalid)}`,
    };
  }

  return { pass: true, score: 1, reason: "skill_file_reads only lists child SKILL.md paths." };
}

export function mutuallyExclusiveSkillsAssert(output, context) {
  const groups = context?.config?.mutually_exclusive_groups ?? [];
  const sections = parseSections(output);
  const actual = new Set(extractSkills(sections.chosen_subskills ?? ""));
  const violations = [];

  for (const group of groups) {
    const present = group.filter((skill) => actual.has(skill));
    if (present.length > 1) {
      violations.push(present);
    }
  }

  if (violations.length) {
    return {
      pass: false,
      score: 0,
      reason: `Found mutually exclusive skills together: ${JSON.stringify(violations)}`,
    };
  }

  return { pass: true, score: 1, reason: "No mutually exclusive skill combination was selected." };
}

export function fastPathBypassAssert(output) {
  const sections = parseSections(output);
  const hasRouting = Object.hasOwn(sections, "chosen_subskills") || Object.hasOwn(sections, "routing_context");

  if (hasRouting) {
    return {
      pass: false,
      score: 0,
      reason: "Trivial task should bypass routing entirely, but found chosen_subskills or routing_context sections.",
    };
  }

  if (!output || !output.trim()) {
    return {
      pass: false,
      score: 0,
      reason: "Trivial task should produce a direct answer, but got empty output.",
    };
  }

  return { pass: true, score: 1, reason: "Trivial task was answered directly without routing overhead." };
}
